<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Genomic Annotation in Livestock for positional candidate LOci • GALLO</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Genomic Annotation in Livestock for positional candidate LOci">
<meta property="og:description" content="Genomic Annotation in Livestock for positional candidate LOci (GALLO) is an R package, for the accurate annotation of genes and Quantitative Trait Loci (QTLs) located within candidate markers and/or regions (haplotypes, windows, CNVs, etc) identified in the most common genomic analyses performed in livestock, such as Genome-Wide Association Studies or transcriptomics. Moreover, GALLO allows the graphical visualization of gene and QTL annotation results, data comparison among different grouping factors (e.g., methods, breeds, tissues, statistical models, studies, etc.), and QTL enrichment in different livestock species including cattle, pigs, sheep, and chicken, among others.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">GALLO</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="articles/GALLO.html">Get started</a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div id="gallo-genomic-annotation-in-livestock-for-positional-candidate-loci" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#gallo-genomic-annotation-in-livestock-for-positional-candidate-loci" class="anchor"></a>GALLO: Genomic Annotation in Livestock for positional candidate LOci</h1></div>
<p>The Genomic Annotation in Livestock for positional candidate LOci (GALLO) is an R package designed to provide a straightforward environment for gene and QTL annotation, as well as data integration from multiple data sources. The QTL enrichment analyses can additionally be performed directly by GALLO using the output obtained from the QTL annotation step. In addition, GALLO also provide a set of functions for graphical visualization for the annotation, comparison, integration and QTL enrichment results. Consequently, GALLO is a useful package for the annotation, identification of hidden pattern across datasets, datamining of previous reported associations, as well as the efficient scrutinization of the genetic architecture of complex traits.</p>
<div id="requirements" class="section level2">
<h2 class="hasAnchor">
<a href="#requirements" class="anchor"></a>Requirements</h2>
<p><em>Depends:</em> R (&gt;= 3.5.0)</p>
<p><em>Imports:</em> rtracklayer, data.table, doParallel, parallel, foreach, lattice, utils, graphics, dplyr, grDevices, boot, RColorBrewer, circlize, dynamicTreeCut, ggplot2, unbalhaar</p>
<p>License: MIT</p>
</div>
<div id="usage" class="section level2">
<h2 class="hasAnchor">
<a href="#usage" class="anchor"></a>Usage</h2>
<ul>
<li>The package starts with the outputs from the most common high-throughput genomic association software such as PLINK, BLUPF90, DESeq2, etc. All that it is necessary in the input file is a column named “CHR”, indicating the chromosome number and a column named “BP” with the chromosomal position in base pairs, when punctual positions are analyzed (i.e., SNPs). On the other hand, when chromosomal windows are evaluated, the column “CHR” is still mandatory, however, now it is necessary the presence of the columns “BP1” and “BP2”. Additionally, the .gtf and .gff files must be provided for gene and QTL annotation respectivelly.</li>
</ul>
<p><em>.gtf files for gene annotation can de found, for example, in ensembl FTP website: <a href="https://www.ensembl.org/info/data/ftp/index.html" class="uri">https://www.ensembl.org/info/data/ftp/index.html</a></em></p>
<p><em>.gtf files for QTL annotation can be found in Animal QTLdb: <a href="https://www.animalgenome.org/QTLdb/" class="uri">https://www.animalgenome.org/QTLdb/</a></em></p>
<ul>
<li>The outputs from each function within GALLO can be used in the downstream functions. For example, the output from find_genes_and_qtls_around_markers() can be used as input of qtl_enrich(), relationship_plot(), etc.</li>
</ul>
<p><strong>To install the package, the following command line can be use in R:</strong></p>
<pre><code><a href="https://devtools.r-lib.org//reference">#install.packages("devtools")
library(devtools)
install_github("pablobio/GALLO")</a></code></pre>
</div>
<div id="functions-description" class="section level2">
<h2 class="hasAnchor">
<a href="#functions-description" class="anchor"></a>Functions description</h2>
<ul>
<li>find_genes_qtls_around_markers: Search genes and QTLs around candidate regions</li>
<li>overlapping_among_groups: Overlapping between grouping factors</li>
<li>plot_overlapping: Plot overlapping between data and grouping factors</li>
<li>plot_qtl_info: Plot QTLs information from the find_genes_qtls_around_markers output</li>
<li>QTLenrich_plot: Plot enrichment results for QTL enrichment analysis</li>
<li>qtl_enrich: Performs a QTL enrichment analysis based in a Bootstrap simulation for each QTL class</li>
<li>relationship_plot: Plot relationship between data and grouping factors</li>
</ul>
<p><strong>Annotation of QTLs overlapping genomic windows, a short example</strong></p>
<pre><code>#Loading package
library(GALLO)

#Loading example dataset
data(QTLwindows)

#Performing QTL annotation (method="qtl") for genomic windows (marker="haplotype"), 
#using an interval of 100Kb upstream and downstream (interval=100000)

qtl.out &lt;- find_genes_qtls_around_markers(db_file="QTL_db.gff",'marker_file=QTLwindows,
method="qtl",'marker="haplotypes",interval=100000)

head(qtl.out)</code></pre>
</div>
<div id="contact" class="section level2">
<h2 class="hasAnchor">
<a href="#contact" class="anchor"></a>Contact</h2>
<p>For more informations, suggestions, dicussions, and bug reports, contact <a href="mailto:pfonseca@uoguelph.ca" class="email">pfonseca@uoguelph.ca</a></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Pablo Fonseca <br><small class="roles"> Author, maintainer </small>  </li>
<li>Aroa Suarez-Vega <br><small class="roles"> Author </small>  </li>
<li>Gabriele Marras <br><small class="roles"> Author </small>  </li>
<li>Angela Cánovas <br><small class="roles"> Author </small>  </li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p>Developed by Pablo Fonseca, Aroa Suarez-Vega, Gabriele Marras, Angela Cánovas.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
